src/core/bpf/allow_bind/allow-bind.c

```diff
@@ -0,0 +1,41 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <bpf/bpf_endian.h>
+#include <linux/bpf.h>
+#include <netinet/in.h>
+
+#include "maps.h"
+
+static inline int allow_bind_impl(
+                struct bpf_sock_addr* ctx,
+                struct allow_ports_t* allowed_ports) {
+        __u16 port_key;
+        __u8* port;
+
+        volatile __u32 user_port = ctx->user_port;
```

Comment:
@poettering 

To prevent a compiler from optimizing `user_port` for narrow load which may be rejected by BPF verifier on older kernels, see a [patch](https://www.spinics.net/lists/netdev/msg532056.html) which describes the issue.

To illustrate how `volatile` affects:

With `volatile`:
```
$ /usr/bin/clang -Wno-compare-distinct-pointer-types -O2 -target bpf -emit-llvm -g -c -D__x86_64__ -I. -idirafter /usr/local/include -idirafter /usr/include ../src/core/bpf/allow_bind/allow-bind.c -o - | \
  /usr/bin/llc -march=bpf -filetype=obj -o ./allow_bind.o
$ # dumping bytecode
$ llvm-objdump -S ./allow_bind.o | less

;         volatile __u32 user_port = ctx->user_port;
       5:       61 11 18 00 00 00 00 00 r1 = *(u32 *)(r1 + 24)
       6:       63 1a f8 ff 00 00 00 00 *(u32 *)(r10 - 8) = r1
;         port_key = (__u16)bpf_ntohs(user_port);
       7:       61 a1 f8 ff 00 00 00 00 r1 = *(u32 *)(r10 - 8)
       8:       dc 01 00 00 10 00 00 00 r1 = be16 r1
       9:       6b 1a fe ff 00 00 00 00 *(u16 *)(r10 - 2) = r1
      10:       bf a2 00 00 00 00 00 00 r2 = r10
      11:       07 02 00 00 fe ff ff ff r2 += -2
```

Without `volatile`:
```
;         __u32 user_port = ctx->user_port;
       5:       61 11 18 00 00 00 00 00 r1 = *(u32 *)(r1 + 24)
       6:       dc 01 00 00 10 00 00 00 r1 = be16 r1
;         port_key = (__u16)bpf_ntohs(user_port);
       7:       6b 1a fe ff 00 00 00 00 *(u16 *)(r10 - 2) = r1
       8:       bf a2 00 00 00 00 00 00 r2 = r10
       9:       07 02 00 00 fe ff ff ff r2 += -2
```

If compiler went further and produce narrow load instruction `r1 = *(u16 *)(r1 + 24)` instead of lines 5-6 there could be a problem.

Reply:
could you add a brief comment about that to the sources? for the next one looking at this